src/rust/generators/sysmon-subgraph-generator/src/generator.rs

```diff
@@ -54,121 +49,76 @@ where
         completed: &mut CompletedEvents,
     ) -> Result<Self::OutputEvent, Result<(Self::OutputEvent, Self::Error), Self::Error>> {
-        info!("Handling raw event");
-
-        /*
-           This iterator is taking a set of bytes of the logs, splitting the logs on newlines,
-           converting the byte sequences to utf-8 strings, and then filtering on the following criteria:
-               1. The line isn't empty
-               2. The line is not `\n` (to prevent issues with multiple newline sequences)
-               3. The line contains event with ID 1, 3, or 11
-
-           The event ids 1, 3, and 11 correspond to Process Creation, Network Connection, and File Creation
-           in that order.
-
-           https://docs.microsoft.com/en-us/sysinternals/downloads/sysmon#events
-        */
-        let events: Vec<_> = log_time!(
-            "event split",
-            events
-                .split(|i| &[*i][..] == &b"\n"[..])
-                .map(String::from_utf8_lossy)
-                .map(|s| s.to_string())
-                .filter(|event| {
-                    (!event.is_empty() && event != "\n")
-                        && (event.contains(&"EventID>1<"[..])
-                            || event.contains(&"EventID>3<"[..])
-                            || event.contains(&"EventID>11<"[..]))
-                })
-                .collect()
-        );
-
-        info!("Handling {} events", events.len());
-
-        let final_subgraph = self.process_events(events, completed).await;
+        info!("Processing {} incoming Sysmon events.", events.len());
+
+        let (deserialized_events, deserialization_errors): (Vec<_>, Vec<_>) =
+            events.into_iter().partition_map(|event| match event {
+                Ok(value) => Either::Left(value),
+                Err(error) => Either::Right(error),
+            });
+
+        for error in deserialization_errors {
+            error!("Deserialization error: {}", error);
+        }
+
+        let subgraph_results = deserialized_events
+            .into_iter()
+            .map(|event| GraphDescription::try_from(event));
+
+        // Report subgraph generation metrics
+        for result in subgraph_results.clone() {
+            self.metrics.report_subgraph_generation(&result);
+        }
+
+        let (subgraphs, mut subgraph_generation_errors): (Vec<_>, Vec<_>) = subgraph_results
+            .partition_map(|result| match result {
+                Ok(value) => Either::Left(value),
+                Err(error) => Either::Right(error),
+            });
+
+        for error in &subgraph_generation_errors {
+            match error {
+                SysmonGeneratorError::UnsupportedEventType(_) => continue,
+                _ => error!("GraphDescription::try_from failed with: {:?}", error),
+            }
+        }
+
+        let final_subgraph =
+            subgraphs
+                .iter()
+                .fold(GraphDescription::new(), |mut current_graph, subgraph| {
```

Comment:
Is there no GraphDescription::merge? It's cool if not, nbd.

Reply:
There is, it's called on the next line, `current_graph.merge(&subgraph);`